File: Data Structures & Algorithms/reorder-linked-list/submission-0.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public void printList(ListNode head) {
         ListNode it = head;

         while(it != null) {
            System.out.print(it.val + " ");
            it= it.next;
        }
        System.out.println("\n");    
    }

    public void reorderList(ListNode head) {
        ListNode reversed = null;
        printList(head);

        ListNode it = head.next;
        int nr = 1;
        int n=0;
        while(it != null) {
            System.out.println("it: " + it.val + " next: " + it.next);
            ListNode tmp = new ListNode(it.val, reversed);
            reversed = tmp;

            it = it.next;
            n++;
        }

        ListNode hIt = head;
        ListNode rIt = reversed;

        ListNode currNode = head;
        for(int i=1; i < n/2; i++) {
            // todo
            hIt.next = rIt;
            hIt = hIt.next;
            rIt = rIt.next;
        }

        printList(head);
    }
}

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-0.kt
/**
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */

class Solution {
    fun mergeLists(arr1: ListNode?, arr2: ListNode?): ListNode? {
        val dummy = ListNode(0)
        var current = dummy

        var pos1 = arr1
        var pos2 = arr2

        while (pos1 != null && pos2 != null) {
            if (pos1.`val` <= pos2.`val`) {
                current.next = pos1
                pos1 = pos1.next
            } else {
                current.next = pos2
                pos2 = pos2.next
            }

            current = current.next!!
        }

        current.next = pos1 ?: pos2

        return dummy.next

    }

    fun mergeKLists(lists: Array<ListNode?>): ListNode? {
        if (lists.isEmpty()) return null

        var result: ListNode? = null

        for (list in lists) {
            result = mergeLists(result, list)
        }

        return result
    }
}